lib/src/tracked_clients/tracked_dio_interceptor.dart

```diff
@@ -0,0 +1,102 @@
+import 'package:appdynamics_agent/appdynamics_agent.dart';
+import 'package:dio/dio.dart';
+
+/// Use this class to create a custom Dio [Interceptor] that tracks requests
+/// automatically. Alternative to [TrackedDioClient].
+///
+/// ```dart
+/// import 'package:dio/dio.dart';
+///
+/// try {
+///   final dio = Dio();
+///   dio.interceptors.add(TrackedDioInterceptor());
+///   final response = await dio.post(urlString, data: {"foo": "bar"});
+///   // handle response
+/// } catch (e) {
+///   // handle error
+/// }
+/// ```
+class TrackedDioInterceptor implements Interceptor {
+  final bool addCorrelationHeaders;
+
+  final Map<String, RequestTracker> _activeTrackers = {};
+
+  TrackedDioInterceptor({this.addCorrelationHeaders = true});
+
+  static const _trackerId = 'trackerId';
+
+  @override
+  void onRequest(
+    RequestOptions options,
+    RequestInterceptorHandler handler,
+  ) async {
+    try {
+      if (addCorrelationHeaders) {
+        final correlationHeaders =
+            await RequestTracker.getServerCorrelationHeaders();
+        final headers = correlationHeaders.map(
+          (key, value) => MapEntry(key, value.first),
+        );
+        options.headers.addAll(headers);
+      }
+
+      var url = options.uri.toString();
+      final tracker = await RequestTracker.create(url);
+      _activeTrackers[tracker.id] = tracker;
+      options.extra[_trackerId] = tracker.id;
+    } finally {
+      handler.next(options);
+    }
+  }
+
+  @override
+  void onResponse(
+    Response<dynamic> response,
+    ResponseInterceptorHandler handler,
+  ) async {
+    try {
+      final tracker =
+          _activeTrackers.remove(response.requestOptions.extra[_trackerId]);
+      if (tracker != null) {
+        await tracker.setResponseStatusCode(response.statusCode ?? 404);
```

Comment:
The fallback should in theory not be needed, as a successful response should always have a status code.
On the other hand, sure, let's keep the fallback if Dio would change their implementation.